gzsitl/gzsitl_plugin.cc

```diff
@@ -130,72 +176,45 @@ void GZSitlPlugin::OnUpdate()
-            model->SetWorldPose(curr_pose);
+        // Wait for take off
+        if (is_flying()) {
+            simstate = ACTIVE_AIRBORNE;
+            print_debug_state("state: ACTIVE_AIRBORNE\n");
         }
+    }
 
-        // Publish current vehicle pose
-        math::Pose mpose_new = model->GetWorldPose();
-        this->vehicle_pose_pub->Publish(msgs::Convert(mpose_new.Ign()));
+    case ACTIVE_AIRBORNE: {
 
-        // Make sure the permanent target still exists to continue
-        perm_target = model->GetWorld()->GetModel(perm_target_name);
-        if (!perm_target) {
-            return;
-        }
+        // Calculate the target azimuthal angle of the target in relation to
+        // the vehicle
+        double targ_ang = (180.0 / M_PI) *
+                          atan2(target_pose_rel.pos.y, target_pose_rel.pos.x);
 
-        // Publish current permanent target pose
-        math::Pose tpose_new = perm_target->GetWorldPose();
-        this->perm_target_pose_pub->Publish(msgs::Convert(tpose_new.Ign()));
+        // Check if the target is located within GZSITL_LOOKAT_TARG_ANG_LIMIT
+        // degrees from the vehicle heading. If not, stop in the current
+        // position and rotate towards target before moving forward.
+        if (fabs(targ_ang) > GZSITL_LOOKAT_TARG_ANG_LIMIT) {
 
-        // Check if target is being overridden by the substitute target
-        if (is_target_overridden()) {
-            tpose_new = get_subs_target_pose();
-        }
+            // Set current target position equal to vehicle position
+            target_pose = vehicle_pose;
 
-        // Set substitute target position in Gazebo if exists
-        subs_target = model->GetWorld()->GetModel(subs_target_name);
-        if (subs_target) {
-            subs_target->SetWorldPose(tpose_new);
+            // Change state on next iteration
+            simstate = ACTIVE_ROTATING;
+            print_debug_state("state: ACTIVE_ROTATING\n");
         }
 
-        // Check if current target position is on the field of view. If not,
-        // stop in the current position and rotate towards target before moving
-        // forward.
-        math::Pose rel_pose = tpose_new - mpose_new;
-
-        double targ_ang = atan2(rel_pose.pos.y, rel_pose.pos.x);
-        targ_ang *= 180.0 / M_PI;
-
-        if (fabs(targ_ang) > GZSITL_LOOKAT_TARG_ANG_LIMIT) {
-            mavserver.queue_send_cmd_long_until_ack(
-                MAV_CMD_CONDITION_YAW, fabs(targ_ang),
-                GZSITL_LOOKAT_ROT_SPEED_DEGPS, -copysign(1, targ_ang), 1, 0, 0,
-                0, 2000);
-            tpose = mpose_new;
-        }
+        // Store static target pose to avoid unnecessary repetition of requests
+        static math::Pose target_pose_prev = math::Pose::Zero;
 
         // Send Target if exists and if it has been moved
-        if (is_flying() && tpose_new != tpose) {
-            tpose = tpose_new;
+        if (is_flying() && target_pose != target_pose_prev) {
```

Comment:
This allows a waypoint to be sent if and only if the vehicle is flying. When the vehicle is on ground, sending the target position is not necessary.